.github/workflows/ci.yaml

```
@@ -8,6 +8,10 @@ on:
 permissions:
   contents: read
 
+concurrency:
+  group: ${{ github.head_ref }}
```

This groups jobs based on pull-request or empty (which means any other non-PR commit would be grouped the same if the action is not PR, more [here](https://docs.github.com/en/actions/using-jobs/using-concurrency#example-using-concurrency-to-cancel-any-in-progress-job-or-run)). This could be an issue for GA running on different action, other workflows, releases, etc. 

This allows a single action to be executed per PR push (via `github.head_ref`) or all actions for a push to main (i.e. commits to main branch will not be grouped since `run_id` is unique to ensure that we don't have red marks on main branch). `github.workflow` groups it by this workflow so that other actions can be executed (that are not). You can check the contexts of GA groups [here](https://docs.github.com/en/actions/learn-github-actions/contexts). 
```suggestion
  group: ${{ github.workflow }}-${{ github.head_ref || github. run_id }}
```

---

I didn't think of different actions at the first place, I appreciate the reminder.